### PR TITLE
python310Packages.mypy: 1.3.0 -> 1.4.1

### PR DESCRIPTION
**Pull request by blaggacao**
###### Description of changes

Changes: https://github.com/python/mypy/compare/v1.3.0...v1.4.1

Required by: https://github.com/NixOS/nixpkgs/pull/243861#issuecomment-1641755257

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @trofi @Artturin

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
1.4.0 release post https://mypy-lang.blogspot.com/2023/06/mypy-140-released.html

**Comment by trofi** [comment]
Does it work for you? For me it fails as:

```
$ nix build --no-link github:NixOS/nixpkgs/pull/244567/merge#python310Packages.mypy -L
...
python3.10-mypy> Executing pipBuildPhase
python3.10-mypy> Creating a wheel...
python3.10-mypy> WARNING: The directory '/homeless-shelter/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
python3.10-mypy> Processing /build/source
python3.10-mypy>   Running command Preparing metadata (pyproject.toml)
python3.10-mypy>   mypy/fastparse.py:223: error: Library stubs not installed for "typed_ast"  [import]
python3.10-mypy>   mypy/fastparse.py:223: note: Error code "import" not covered by "type: ignore" comment
python3.10-mypy>   mypy/fastparse.py:223: note: Hint: "python3 -m pip install types-typed-ast"
python3.10-mypy>   mypy/fastparse.py:223: note: (or run "mypy --install-types" to install all missing stub packages)
python3.10-mypy>   mypy/fastparse.py:223: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
python3.10-mypy>   error: subprocess-exited-with-error
python3.10-mypy>   × Preparing metadata (pyproject.toml) did not run successfully.
python3.10-mypy>   │ exit code: 1
python3.10-mypy>   ╰─> See above for output.
python3.10-mypy>   note: This error originates from a subprocess, and is likely not a problem with pip.
python3.10-mypy>   full command: /nix/store/a282mi66s8msrkn67kn0zbycc2kajdhp-python3-3.10.12/bin/python3.10 /nix/store/rn82klzdnnkzh5lm42zy05mp0bv63fj5-python3.10-pip-23.0.1/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmpqmxy40lu
python3.10-mypy>   cwd: /build/source
python3.10-mypy>   Preparing metadata (pyproject.toml) ... error
python3.10-mypy> error: metadata-generation-failed
python3.10-mypy> × Encountered error while generating package metadata.
python3.10-mypy> ╰─> See above for output.
python3.10-mypy> note: This is an issue with the package mentioned above, not pip.
python3.10-mypy> hint: See above for details.
error: builder for '/nix/store/s81wjrf74m4wyd6zrawmfflz5dhdf5z7-python3.10-mypy-1.4.1.drv' failed with exit code 1;
```

**Comment by blaggacao** [comment]
Unfortunately, I have no cache of `staging`, atm.

**Comment by Artturin** [comment]
try updating `types-typed-ast` to `1.5.8.7 `

File: pkgs/development/python-modules/mypy/default.nix [hunk 1]
```diff
@@ -31,24 +31,23 @@
...
-    hash = "sha256-dfKuIyzgZo5hAZHighpXH78dHJ1PMbyCakyxF34CnMQ=";
+    hash = "sha256-2PeE/L9J6J0IuUpHZasemM8xxefNJrdzYnutgJjevWQ=";
   };
 
   nativeBuildInputs = [
     mypy-extensions
     setuptools
     types-psutil
     types-setuptools
-    types-typed-ast
```
Comment (blaggacao) [inline pkgs/development/python-modules/mypy/default.nix:51]: I think this was a mistake. :smile:

Comment (blaggacao) [reply pkgs/development/python-modules/mypy/default.nix:51]: Sure it was. It was only remove on main branch, but not on 1.4.1

**Comment by trofi** [comment]
Retried today:

```
$ nix build --no-link github:NixOS/nixpkgs/pull/244567/merge#python310Packages.mypy -L
...
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNoSitePkgsIgnoredImports
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgSimple
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgSimplePackageSearchPath
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkg_config_nositepackages
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkg_args_nositepackages
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgStubs
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testStubPrecedence
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgSimpleEditable
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceImportFrom
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceImportAs
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceRegImport
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testPep561TestIncremental
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceRegFromImportTwice
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testNamespacePkgWStubs
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testNamespacePkgWStubsWithNamespacePackagesFlag
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceRegFromImportTwiceMissing
python3.10-mypy> FAILED mypy/test/testpep561.py::PEP561Suite::pep561.test::testTypedPkgNamespaceRegFromImportTwiceMissing2
python3.10-mypy> ==== 17 failed, 10862 passed, 332 skipped, 10 xfailed in 268.25s (0:04:28) =====
```

**Comment by blaggacao** [comment]
Unfortunately, for reasons I don't quite understand, the following diff does not afford [the required `hatchling` for tests](https://github.com/python/mypy/blob/master/test-data/packages/typedpkg_ns_a/pyproject.toml#L10):

<details><summary>diff</summary>

```diff
diff --git a/pkgs/development/python-modules/mypy/default.nix b/pkgs/development/python-modules/mypy/default.nix
index a566d4b5a7f..384e7cf036d 100644
--- a/pkgs/development/python-modules/mypy/default.nix
+++ b/pkgs/development/python-modules/mypy/default.nix
@@ -22,6 +22,7 @@
 # tests
 , attrs
 , filelock
+, hatchling
 , pytest-xdist
 , pytest-forked
 , pytestCheckHook
@@ -90,9 +91,10 @@ buildPythonPackage rec {
     "mypy.report"
   ];insights
 
-  checkInputs = [
+  nativeCheckInputs = [
     attrs
     filelock
+    hatchling
     pytest-xdist
     pytest-forked
     pytestCheckHook
```

</details>

**Comment by mweinelt** [comment]
Updated in fb198ee98918fde93352fe4aefeac1a71ff93e43.

As with setuptools and typing_extensions it also does not find hatchling. Nothing t worry about. I think it spawns a new environment and tries to install it from pypi.

**Pull request closed without merging**